### ISSUE TITLE
featureCounts name-sorts BAM input file in current working directory with a single core

### ISSUE THREAD

**Issue by schelhorn** [chapmanb/bcbio-nextgen#1013]
Usually, in the `rnaseq` workflow, quantitation for paired inputs with `featureCounts` runs on coordinate-sorted BAM inputs (since, for instance `cufflinks` requires coordinate-sorted inputs). Due to internal requirements of `featureCounts` (ie., since it only counts feature hits if both read ends align to the same feature), these coordinate-sorted BAM input files are detected by the `featureCounts` binary and name-sorted prior to processing. Name-sorting occurs by `featureCounts` itself (as part of the binary), on a single core and in the current working directory. 

As a consequence, that sorting happens in the project working directory of bcbio that usually resides on a shared file system. As a result, a lot of single-core processing power is wasted on sorting these BAMs and the shared file system breaks down if you run a couple of hundred samples in parallel. 

It would be great if instead the bcbio featurecounts.py module could name-sort the inputs before calling `featureCounts` using the appropriate bcbio function so that sorting happens on local disk and with the specified number of `samtools` cores. `featureCounts` would then (hopefully) refrain from sorting it again. 

Note that the name-sorted BAM file is only used by `featureCounts` anyway (as far as I know), so that file could stay on local disk and be removed after `featureCounts` finishes. In particular, there is no need to copy it over to the shared file system at all (technically, sorting could just be run in the `featureCounts` execution context on the same node `featureCounts` is ran, but one would require `samtools` to be in the parallel environment resource specs, of course).

**Comment by mjafin** [comment]
As an aside, didn't STAR now provide counts directly and thus we could save some time and drop `featureCounts` for when STAR is used. Any flaws in my line of thinking?

**Comment by schelhorn** [comment]
@mjafin, true and certainly preferred from my end, although one could argue that the additional filtering options available to `featureCounts` (such as only counting when both mate ends align to the feature) are also nice to have. Still, I'd prefer the STAR way here, too. Also, there is a second `STAR` `--quantMode` option, `TranscriptomeSAM`. Both quantitations can be run in parallel, so we could also do away with the separate `bowtie2` transcriptome alignment for `express` (also allows for direct BAM output and RSEM compatibility). @roryk, you're managing the `rnaseq` workflow, what is your thinking in this regard?

**Comment by roryk** [comment]
Hi @schelhorn and @mjafin,

featureCounts is a little bit finicky about how the read pairs are sorted and some people have reported it can fail on queryname sorted files from samtools and the unsorted output of STAR (see for example http://seqanswers.com/forums/showthread.php?t=49687).  

I've been reluctant about using the STAR-generated counts because featureCounts is pretty fast once the file is sorted and it is nice to have the counting be the same regardless of the aligner being used and we'll run into this same problem with Tophat2 as well. 

I added sorting by queryname with sambamba before passing it into featureCounts, but that may not fix the issue. I posted upstream too to see if featureCounts could get fixed to handle some of the edge cases in the queryname sorted files by itself: https://groups.google.com/forum/#!topic/subread/ZYBhkybL904

**Comment by roryk** [comment]
We could clean the file ourselves before passing it in if it looks like just presorting it doesn't stop featureCounts from sorting it at all.

**Comment by schelhorn** [comment]
Thanks for the fix, Rory! Regarding the format required by featureCounts, I believe we could get increased compatibility by only retaining primary alignments in the name-sorted input files. I'm not sure if the aligners are configured to only specify one alignment as primary, but if not then we could fix that with Picard I believe (hopefully in a streaming fashion). As a result, mate pairs should be next to each other in the BAM file since each mate only had one alignment. Usually quantification only considered the best alignment anyway, so we shouldn't loose anything doing this.

**Comment by roryk** [comment]
Agreed-- STAR marks the best or a random best alignment as primary if the alignments are equal and I think Tophat2 does the same thing if `--report-secondary-alignments` isn't set so just filtering on that flag might do the trick.

**Comment by roryk** [comment]
Hi @schelhorn and @mjafin,

I fixed this here: ef1ddf17cf5c22. featureCounts unfortunately can't handle streaming so it has to go to a temporary file.

It might be that newer versions of samtools or sambamba will put the reads in the right order all of the time, I'm not really sure how to figure out if that is the case or not. An alternative would be just to skip this step and see if we start having problems, but I think featureCounts can end up just silently doing the wrong thing.

**Comment by schelhorn** [comment]
Thank you, excellent work. Closing.
